### PR TITLE
feat: lambda support for yarn2/3 and pnpm package managers

### PR DESCRIPTION
**Pull request by lazpavel**
#### Description of changes
- adds support for PNPM and Yarn 2 package managers in lambda functions
- adds support for custom command or scripts to build lambda functions
- adds support for PNPM and Yarn 2 package managers for custom resouces

#### Description of how you validated changes
- manual testing and e2e test added

#### Checklist
- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-e2e-tests/src/__tests__/function_13.test.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ describe('amplify push function cases:', () => {
     deleteProjectDir(projRoot);
   });
 
-  it('should select the NPM package manager and build', async () => {
+  it.only('should select the NPM package manager and build', async () => {
```
Comment (jhockett) [inline packages/amplify-e2e-tests/src/__tests__/function_13.test.ts]: debugging leftover?

File: packages/amplify-cli-core/API.md [hunk 2]
```diff
@@ -2131,6 +2139,12 @@ export interface Template {
 // @public (undocumented)
 export type TimedCodePath = ManuallyTimedCodePath | UntilExitTimedCodePath | FromStartupTimedCodePaths;
 
+// @public (undocumented)
+export const toPackageManagerInstallArgs: (packageManager: PackageManager, buildType?: BuildType) => Promise<string[]>;
+
+// @public (undocumented)
+export const toPackageManagerRunScriptArgs: (packageManager: PackageManager, scriptName: string) => Promise<string[]>;
```
Comment (edwardfoyle) [inline packages/amplify-cli-core/API.md:2146]: should these be methods on a PackageManager class?

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 3]
```diff
@@ -72,12 +90,51 @@ export const getPackageManager = async (rootPath?: string): Promise<PackageManag
...
+      return ['run-script', scriptName];
+    default: {
+      throw new AmplifyError('PackagingLambdaFunctionError', {
+        message: `Packaging lambda function failed. Unsupported package manager ${packageManager.packageManager}`,
+      });
+    }
+  }
+};
+
+export const toPackageManagerInstallArgs = async (packageManager: PackageManager, buildType = BuildType.PROD): Promise<string[]> => {
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts:118]: should not be anymore

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 4]
```diff
@@ -72,12 +90,51 @@ export const getPackageManager = async (rootPath?: string): Promise<PackageManag
...
+  }
+
+  if (checkExecutable(packageManagers.pnpm.executable)) {
+    return packageManagers.pnpm;
   }
 
   return packageManagers.npm;
 };
 
+export const toPackageManagerRunScriptArgs = async (packageManager: PackageManager, scriptName: string): Promise<string[]> => {
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts:103]: not needed anymore, removed

File: .eslint-dictionary.json [hunk 5]
```diff
@@ -340,6 +341,7 @@
   "sdk",
   "sdks",
   "searchable",
+  "secres",
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply .eslint-dictionary.json:344]: removed

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts]: converted to interface and extended classes

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/functionPluginLoader.ts [hunk 6]
```diff
@@ -83,18 +93,42 @@ export async function runtimeWalkthrough(
...
+      value: pm.packageManager as string,
+    }));
+
+    packageManagerOptions.push({
+      name: 'Custom Build Command or Script Path',
+      value: 'custom',
+    });
+
+    const packageManager = (await prompter.pick('Choose the package manager that you want to use:', packageManagerOptions)) as
+      | PackageManagerType
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/functionPluginLoader.ts]: bundled under packageManagers

File: packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/packageManagerWalkthrough.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  if (runtime === 'nodejs') {
+    const packageManagerOptions = Object.values(packageManagers).map((pm) => ({
+      name: pm.displayValue,
+      value: pm.packageManager as string,
+    }));
+
+    const packageManager = (await prompter.pick(
+      'Choose the package manager that you want to use:',
+      packageManagerOptions,
+    )) as PackageManagerType;
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/packageManagerWalkthrough.ts:15]: I tried this, the compiler was complaining that is why I had to do this as string/as PackageManagerType thing, maybe I am just doing something wrong

File: packages/amplify-category-function/package.json [hunk 8]
```diff
@@ -47,7 +47,8 @@
     "uuid": "^8.3.2"
   },
   "devDependencies": {
-    "@types/folder-hash": "^4.0.1"
+    "@types/folder-hash": "^4.0.1",
+    "jest": "^29.5.0"
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-function/package.json:51]: yes, since yarn3 dependencies used for each package needs to be explicitly defined, otherwise running yarn test withing the package will fail

File: packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/packageManagerWalkthrough.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/packageManagerWalkthrough.ts:15]: Argument of type '{ name: string; value: PackageManagerType; }[]' is not assignable to parameter of type 'string[] | GenericChoice<"yarn">[] | GenericChoice<"npm">[] | GenericChoice<"pnpm">[] | GenericChoice<"custom">[]'.

File: packages/amplify-cli-core/package.json [hunk 9]
```diff
@@ -29,6 +29,7 @@
   },
   "dependencies": {
     "@aws-amplify/amplify-cli-logger": "1.3.3",
+    "@aws-amplify/amplify-function-plugin-interface": "1.10.3",
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/package.json:32]: no, this package does not depend on anything

File: packages/amplify-cli/package.json [hunk 10]
```diff
@@ -121,6 +121,7 @@
     "@types/treeify": "^1.0.0",
     "@types/update-notifier": "^5.1.0",
     "cloudform-types": "^4.2.0",
+    "jest": "^29.5.0",
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli/package.json:124]: yes, this is needed if you want to run `yarn` scripts within the package directory

File: packages/amplify-cli-core/src/utils/packageManager.ts
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts:1]: discussed offline

File: packages/amplify-cli-core/src/amplify-node-pkg-detector/parser-factory.ts [hunk 11]
```diff
@@ -26,7 +26,8 @@ export class LockFileParserFactory {
       }
       default:
         throw new AmplifyError('UnsupportedLockFileTypeError', {
-          message: 'Unsupported lockfile type ' + `${packageManager.lockFile} provided. Only 'npm' or 'yarn' is currently ` + 'supported.',
+          message:
+            'Unsupported lockfile type ' + `${packageManager.lockFile} provided. Only 'npm', 'yarn', 'pnpm' is currently ` + 'supported.',
```
Comment (jhockett) [inline packages/amplify-cli-core/src/amplify-node-pkg-detector/parser-factory.ts:30]:
```suggestion
            `Unsupported lockfile type ${packageManager.lockFile} provided. Only 'npm', 'yarn', and 'pnpm' are currently supported.`,
```

**Review by jhockett — DISMISSED** [review]

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 12]
```diff
@@ -1,39 +1,115 @@
...
 import { execWithOutputAsString } from './shell-utils';
+import { AmplifyError } from '../errors/amplify-error';
+import { BuildType } from '@aws-amplify/amplify-function-plugin-interface';
 
 /**
  * package managers type
  */
-export type PackageManagerType = 'yarn' | 'npm';
-
+export type PackageManagerType = 'yarn' | 'npm' | 'pnpm' | 'custom';
```
Comment (Amplifiyer) [inline packages/amplify-cli-core/src/utils/packageManager.ts:12]: Isn't this package manager also used for compiling overrides/custom resources? Will this PR enable compiling the overrides/custom as well?

File: packages/amplify-cli-core/src/amplify-node-pkg-detector/parser-factory.ts [hunk 13]
```diff
@@ -26,7 +26,7 @@ export class LockFileParserFactory {
       }
       default:
         throw new AmplifyError('UnsupportedLockFileTypeError', {
-          message: 'Unsupported lockfile type ' + `${packageManager.lockFile} provided. Only 'npm' or 'yarn' is currently ` + 'supported.',
+          message: `Unsupported lockfile type ${packageManager.lockFile} provided. Only 'npm', 'yarn', and 'pnpm' are currently supported.`,
```
Comment (Amplifiyer) [inline packages/amplify-cli-core/src/amplify-node-pkg-detector/parser-factory.ts:29]: In this file above we have `LockFileType` which current has Yarn or NPM, do we not need PNPM here as well? and Yarn 3?

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 12 above)
Comment (Amplifiyer) [reply packages/amplify-cli-core/src/utils/packageManager.ts:12]: cc: @akshbhu

File: packages/amplify-cli-core/src/amplify-node-pkg-detector/parser-factory.ts (hunk 13 above)
Comment (lazpavel) [reply packages/amplify-cli-core/src/amplify-node-pkg-detector/parser-factory.ts:29]: this parser-factory logic ends up being used only here, for a migration message: https://github.com/lazpavel/amplify-cli/blob/lambda-package-manager/packages/amplify-provider-awscloudformation/src/print-cdk-migration-warning.ts#L106-L133

since PNPM is newly added we don't currently need it being added here

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 12 above)
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts:12]: this seems to be the case, I was not aware of.

https://github.com/lazpavel/amplify-cli/blob/lambda-package-manager/packages/amplify-category-custom/src/utils/build-custom-resources.ts#L71-L94

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 14]
```diff
@@ -1,39 +1,115 @@
...
 const packageJson = 'package.json';
 
 /**
  * package Manager type
  */
-export type PackageManager = {
+export interface PackageManager {
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts:18]: discussed offline, agreed to leave as interface

File: packages/amplify-e2e-core/src/categories/api.ts [hunk 15]
```diff
@@ -253,6 +253,8 @@ export function addApiWithAllAuthModes(cwd: string, opts: Partial<AddApiOptions
       // Lambda
       .wait(/.*Choose a Lambda authorization function*/)
       .sendCarriageReturn()
+      .wait('Choose the package manager that you want to use')
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-e2e-core/src/categories/api.ts]: moved under advanced settings

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts]: leftover, removed

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/functionPluginLoader.ts [hunk 16]
```diff
@@ -83,18 +93,42 @@ export async function runtimeWalkthrough(
...
 }
 
+const packageManagerWalkthrough = async (selection: PluginSelection): Promise<void> => {
+  if (selection.value === 'nodejs') {
+    const packageManagerOptions = Object.values(packageManagers).map((pm) => ({
+      name: pm.displayValue,
+      value: pm.packageManager as string,
+    }));
+
+    packageManagerOptions.push({
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/functionPluginLoader.ts]: added to the list

File: packages/amplify-cli-core/API.md (hunk 2 above)
Comment (lazpavel) [reply packages/amplify-cli-core/API.md]: converted to class methods

**Review by jhockett — DISMISSED** [review]

**Review by jhockett — DISMISSED** [review]

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 12 above)
Comment (akshbhu) [reply packages/amplify-cli-core/src/utils/packageManager.ts:12]: This seems like it can break on cli if Cx chooses ```custom``` type  for a function.

Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts:12]: if the Cx chooses custom for a function, only that function will be built with `custom` not other functions or custom resources, for custom resources the prompt is not presented and the decision will be made based on the custom resource directory config, which lock files exist, if non exist, it will pick first available executable in the system: yarn, pnpm, npm in this order of testing, also npm is used regardless if it exists or not in the system (old behavior)

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 17]
```diff
@@ -9,41 +9,103 @@ import { BuildType } from '@aws-amplify/amplify-function-plugin-interface';
...
+export class NpmPackageManager implements PackageManager {
+  readonly packageManager: PackageManagerType = 'npm';
+  readonly displayValue = 'NPM';
+  lockFile;
+  executable;
+  version?: SemVer;
+
+  constructor() {
+    this.lockFile = 'package-lock.json';
+    this.executable = isWindows ? 'npm.cmd' : 'npm';
```
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/utils/packageManager.ts:39]: seems like these could be defined as readonly values along with packageManager and displayValue (same for other classes)

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 18]
```diff
@@ -9,41 +9,103 @@ import { BuildType } from '@aws-amplify/amplify-function-plugin-interface';
...
+    });
+  };
+  getInstallArgs = () => {
+    throw new AmplifyError('PackagingLambdaFunctionError', {
+      message: `Packaging lambda function failed. Unsupported package manager`,
+    });
+  };
+}
+
 export const packageManagers: Record<string, PackageManager> = {
```
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/utils/packageManager.ts:104]:
```suggestion
export const packageManagers: Record<PackageManagerType, PackageManager> = {
```

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 19]
```diff
@@ -9,41 +9,103 @@ import { BuildType } from '@aws-amplify/amplify-function-plugin-interface';
...
+}
 
 const isWindows = process.platform === 'win32';
 
+export class NpmPackageManager implements PackageManager {
+  readonly packageManager: PackageManagerType = 'npm';
+  readonly displayValue = 'NPM';
+  lockFile;
+  executable;
+  version?: SemVer;
```
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/utils/packageManager.ts:35]: if this isn't used, remove it

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 20]
```diff
@@ -9,41 +9,103 @@ import { BuildType } from '@aws-amplify/amplify-function-plugin-interface';
...
   version?: SemVer;
-  displayValue?: string;
-};
+  getRunScriptArgs: (scriptName: string) => string[];
+  getInstallArgs: (buildType: BuildType) => string[];
+}
 
 const isWindows = process.platform === 'win32';
 
+export class NpmPackageManager implements PackageManager {
```
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/utils/packageManager.ts:30]: should these classes be internal to this package so they have to be retrieved using `getPackageManagerByType`?

File: packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts [hunk 21]
```diff
@@ -61,7 +53,7 @@ const runPackageManagerInstall = async (resourceDir: string, buildType: BuildTyp
     return;
   }
 
-  const args = await toPackageManagerInstallArgs(packageManager, buildType);
+  const args = await packageManager.getInstallArgs(buildType);
```
Comment (edwardfoyle) [inline packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts:56]: unnecessary await

File: packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts [hunk 22]
```diff
@@ -73,7 +65,7 @@ const runPackageManagerScript = async (resourceDir: string, scriptName: string):
     return;
   }
 
-  const args = await toPackageManagerRunScriptArgs(packageManager, scriptName);
+  const args = await packageManager.getRunScriptArgs(scriptName);
```
Comment (edwardfoyle) [inline packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts:68]: unnecessary await

File: packages/amplify-cli/src/__tests__/commands/init.test.ts [hunk 23]
```diff
@@ -20,14 +37,12 @@ jest.mock('which');
...
 (getPackageManager as jest.MockedFunction<typeof getPackageManager>).mockReturnValue(
-  new Promise((resolve) =>
-    resolve({
-      executable: 'yarn',
-      lockFile: 'mock.lock',
-      packageManager: 'yarn',
-      version: coerce('1.22.0') ?? undefined,
-    }),
-  ),
+  new Promise((resolve) => {
```
Comment (edwardfoyle) [inline packages/amplify-cli/src/__tests__/commands/init.test.ts:40]: nit: use `mockResolvedValue` without the `new Promise` wrapper

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 20 above)
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts:30]: will do

File: packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts (hunk 21 above)
Comment (lazpavel) [reply packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts:56]: removed

File: packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts (hunk 22 above)
Comment (lazpavel) [reply packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts:68]: removed

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 17 above)
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts]: updated

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 18 above)
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts]: updated

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 24]
```diff
@@ -9,41 +9,103 @@ import { BuildType } from '@aws-amplify/amplify-function-plugin-interface';
...
+  readonly packageManager: PackageManagerType = 'npm';
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts]: yes, removed the type

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 19 above)
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts]: removed the unused ones

File: packages/amplify-cli/src/__tests__/commands/init.test.ts (hunk 23 above)
Comment (lazpavel) [reply packages/amplify-cli/src/__tests__/commands/init.test.ts]: updated

File: packages/amplify-cli-core/src/utils/packageManager.ts [hunk 25]
```diff
@@ -27,58 +27,41 @@ export interface PackageManager {
...
-  lockFile;
-  executable;
-  version?: SemVer;
-
-  constructor() {
-    this.lockFile = 'package-lock.json';
-    this.executable = isWindows ? 'npm.cmd' : 'npm';
-  }
+  readonly executable = isWindows ? 'npm.cmd' : 'npm';
+  readonly lockFile = 'package-lock.json';
```
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/utils/packageManager.ts]: these props need to be readonly in the interface as well so TS will enforce it in other places

File: packages/amplify-cli/src/__tests__/commands/init.test.ts [hunk 26]
```diff
@@ -36,14 +36,11 @@ jest.mock('which');
...
-  new Promise((resolve) => {
-    const packageManager = new YarnPackageManager();
-    packageManager.lockFile = 'mock.lock';
-    packageManager.version = coerce('1.22.0') ?? undefined;
-    resolve(packageManager);
-  }),
-);
+
+const packageManager = getPackageManagerByType('yarn');
+packageManager.lockFile = 'mock.lock';
```
Comment (edwardfoyle) [inline packages/amplify-cli/src/__tests__/commands/init.test.ts:41]: Once you make `lockFile` readonly on the interface you won't be able to do this, but I think that's okay

**Review by edwardfoyle — COMMENTED** [review]
One small update, then I think this is g2g

File: packages/amplify-cli/src/__tests__/commands/init.test.ts (hunk 26 above)
Comment (lazpavel) [reply packages/amplify-cli/src/__tests__/commands/init.test.ts:41]: that is the only reason I didn't make it readonly on the interface, if we are good to introduce as $TSAny, for me to keep the tests running then I can update

File: packages/amplify-cli-core/src/utils/packageManager.ts (hunk 25 above)
Comment (lazpavel) [reply packages/amplify-cli-core/src/utils/packageManager.ts]: updated

File: packages/amplify-cli-core/src/__tests__/amplify-lockfile-dependency-detector/amplify-nodejs-detector.test.ts [hunk 27]
```diff
@@ -26,7 +27,7 @@ const npmPackageManager = getPackageManagerByType('npm');
 
 describe('parsing yarn lock files', () => {
   it('throws error when lock file not found', async () => {
-    yarnPackageManager.lockFile = 'yarn-test-not-found.lock';
+    (yarnPackageManager as $TSAny).lockFile = 'yarn-test-not-found.lock';
```
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/__tests__/amplify-lockfile-dependency-detector/amplify-nodejs-detector.test.ts:30]: Seems like there's probably a way to updates these tests to check the same behavior without overriding this value, but I'm okay if that is out of scope for now

**Review by edwardfoyle — APPROVED** [review]

File: packages/amplify-provider-awscloudformation/package.json [hunk 28]
```diff
@@ -79,7 +79,8 @@
     "@types/lodash.throttle": "^4.1.6",
     "@types/node": "^12.12.6",
     "@types/uuid": "^8.0.0",
-    "jest": "^29.5.0"
+    "jest": "^29.5.0",
+    "typescript": "^5.1.3"
```
Comment (jhockett) [inline packages/amplify-provider-awscloudformation/package.json:83]: We already depend on `4.9.5` in the root package.json. Is this needed? If so, can we use `4.9.5` instead?

**Review by jhockett — APPROVED** [review]

File: packages/amplify-provider-awscloudformation/package.json (hunk 28 above)
Comment (lazpavel) [reply packages/amplify-provider-awscloudformation/package.json]: we need to add it to each package if you want to execute yarn commands that call tsc from within the package directory, updated the version to match the rest of the project

**Review by edwardfoyle — APPROVED** [review]
